AGW: s1-integ-test: validate HE envoy config.

pshelar opened the pull request:
Signed-off-by: Pravin B Shelar <pbshelar@fb.com>

<!--
    Tag your PR title with the components that it touches.
    E.g. "[lte][agw] Changeset" or "[orc8r][docker] ..."
-->

## Summary

<!-- Enumerate changes you made and why you made them -->

## Test Plan

<!--
    How did you test your change? How do you know it works?
    Add supporting screenshots, terminal pastes, etc. as necessary
-->

## Additional Information

- [ ] This change is backwards-breaking

<!--
    If this is a backwards-breaking change, document the upgrade instructions.
    All upgrade instructions for backwards-breaking changes will be aggregated
    in the next release's changelog so this is very important to fill out.
-->

koolzz commented on lte/gateway/python/integ_tests/s1aptests/test_attach_he_policy.py:
```diff
@@ -0,0 +1,239 @@
+
+        # Verify if paging flow rules are created
+        ip_list = [default_ip, sec_ip]
+        self._s1ap_wrapper.s1_util.verify_paging_flow_rules(ip_list)
```
Do we need to do this in this test? Here and in a few other places we're verifying quite a few non he things

pshelar replied:
I kept the validations that are copied from other tests to confirm additional HE parameter in rule does not change existing flows and attach procedure.

ardzoht commented on lte/gateway/configs/pipelined.yml:
```diff
@@ -160,6 +160,7 @@ uplink_gw_mac: 'ff:ff:ff:ff:ff:ff'
 ovs_gtp_stats_polling_interval: 60
 
 he_proxy_eth_mac: 'e6:8f:a2:80:80:80'
+he_enabled: true
```
should this also be added on `pipelined_prod.yml`?

pshelar replied:
This is enabled via mconfig on 1.4, I will add it for 1.5 by default.

koolzz commented on lte/gateway/python/integ_tests/s1aptests/s1ap_utils.py:
```diff
@@ -1322,3 +1325,53 @@ def offload_agw(self, imsi, enbID, offloadtype=0):
+
+        for conf in self.dump['configs']:
+            if 'static_route_configs' in conf:
+                return conf['static_route_configs']
```
```
        for conf in self.dump['configs']:
            if 'dynamic_listeners' in conf:
                return conf['dynamic_listeners'][0]['active_state']['listener']['filter_chains'][0]['filters']
```

pshelar replied:
I see, I did not notice the 'dynamic_listener' section. I think that is right config to set.
But we do need to check why 'static_coute_config' is set in envoy-config.

koolzz commented on lte/gateway/python/integ_tests/s1aptests/s1ap_utils.py:
```diff
@@ -1322,3 +1325,53 @@ def offload_agw(self, imsi, enbID, offloadtype=0):
+        envoy_conf1 = self.get_route_config()
+        cnt = 0
+        for conf in envoy_conf1:
+            virtual_host_config = conf['route_config']['virtual_hosts']
```
```
virtual_host_config = conf['typed_config']['route_config']['virtual_hosts']
```

koolzz reviewed (changes requested):
I think you're comparing the wrong routing configuration, I've changed the lookup and now the test passes(see code comments). 
Currently we don't even set the snapshot route(the one that you're currently comparing), so I'm not sure how it gets set. We only set the active listners/clusters (`snap := cache.NewSnapshot(fmt.Sprint(cli.version), nil, cluster, nil, listener_resource, nil))`
I don't fully understand whats the purpose of the route when we set all the required information in the listeners, lmk if we need to change something on the controller side.

ulaskozat commented on lte/gateway/python/integ_tests/s1aptests/test_attach_he_policy.py:
```diff
@@ -0,0 +1,232 @@
+from lte.protos.policydb_pb2 import FlowMatch, HeaderEnrichment
+
+
+class TestAttachServiceWithMultiPdnsAndBearers(unittest.TestCase):
```
change the test name in line with the file name.

ulaskozat replied:
what is the reason that you prefer to base the test on top of multi pdn and bearer testing?

pshelar replied:
I changed base to test_attach_detach_rar_tcp_data.py.

ulaskozat commented:
could you fill in the summary section and testing?

ulaskozat commented on lte/gateway/python/integ_tests/s1aptests/s1ap_utils.py:
```diff
@@ -472,7 +474,7 @@ def verify_flow_rules(self, num_ul_flows, dl_flow_rules=None):
                 break
             time.sleep(5)  # sleep for 5 seconds before retrying
-        assert len(uplink_flows) == num_ul_flows, "Uplink flow missing for UE"
+        assert len(uplink_flows) == num_ul_flows, "Uplink flow missing for UE %d" % len(uplink_flows)
```
this print will lead to a wrong impression that len(uplink_flows) is the UE id.

pshelar replied:
fixed

ulaskozat commented on lte/gateway/python/integ_tests/s1aptests/test_attach_he_policy.py:
```diff
@@ -0,0 +1,232 @@
+
+    def test_attach_service_with_he_policy(self):
+        """
+        This test validates the pipelineD is pushing currect config
```
typo "currect" -> "the correct"
validates the -> validates that

ulaskozat commented on lte/gateway/python/integ_tests/s1aptests/test_attach_he_policy.py:
```diff
@@ -0,0 +1,232 @@
+        imsi = "IMSI" + "".join([str(i) for i in req.imsi])
+        self._s1ap_wrapper.configAPN(
+            "IMSI" + "".join([str(i) for i in req.imsi]), apn_list
+        )
```
do you really need a second APN for this test case?

pshelar replied:
changed it to single APN test

ulaskozat commented on lte/gateway/python/integ_tests/s1aptests/test_attach_he_policy.py:
```diff
@@ -0,0 +1,232 @@
+        )
+        act_def_bearer_req = response.cast(s1ap_types.uePdnConRsp_t)
+        addr = act_def_bearer_req.m.pdnInfo.pAddr.addrInfo
+        sec_ip = ipaddress.ip_address(bytes(addr[:4]))
```
see my comment regarding the secondary PDN. these lines will not be needed if you remove the second PDN in the test.

ulaskozat commented on lte/gateway/python/integ_tests/s1aptests/test_attach_he_policy.py:
```diff
@@ -0,0 +1,232 @@
+            "********************** Sending Activate default EPS bearer "
+            "context accept for UE id ",
+            ue_id,
+        )
```
this can be also removed...

ulaskozat commented on lte/gateway/python/integ_tests/s1aptests/test_attach_he_policy.py:
```diff
@@ -0,0 +1,232 @@
+            default_ip: [flow_list1],
+        }
+        # 1 UL flow is created per bearer
+        num_ul_flows = 2
```
I am surprised here that the test is passing in this form. From what I see there should be two sessions with first session  having one default+one dedicated and the second session having one default bearer. So UL flow count should be 3 unless you remove the second PDN.

ulaskozat commented on lte/gateway/python/integ_tests/s1aptests/test_attach_he_policy.py:
```diff
@@ -0,0 +1,232 @@
+
+        # Verify if paging flow rules are created
+        ip_list = [default_ip, sec_ip]
+        self._s1ap_wrapper.s1_util.verify_paging_flow_rules(ip_list)
```
no need to include idle/active mode transitions in this test.

pshelar replied:
active/inactive does not make difference to the HE policy config.

ulaskozat commented on lte/gateway/python/integ_tests/s1aptests/test_attach_he_policy.py:
```diff
@@ -0,0 +1,232 @@
+        self._s1ap_wrapper.s1_util.verify_flow_rules(
+            num_ul_flows, dl_flow_rules
+        )
+
```
Lines 193-217 can be removed if idle/active mode transition is to be eliminated. for this test, looks redundant...

ulaskozat reviewed (changes requested):
see my comments.

119Vik commented:
RestartCWAGTests

pshelar committed:
AGW: HE: integ test

This test setup HE policy rule and validates config
in the enoy proxy.
Second test validates config as well as flows created in
OVS flow table. this test is targeted for pre-commit test
list

Signed-off-by: Pravin B Shelar <pbshelar@fb.com>

ulaskozat reviewed (approved):
LGTM

ardzoht commented on lte/gateway/python/integ_tests/s1aptests/s1ap_utils.py:
```diff
@@ -1508,6 +1513,20 @@ def get_gtp_port_no(self) -> Optional[int]:
+        for line in output.split('\n'):
+            if self.proxy_port in line:
+                port_info = line.split()
+                return port_info[1]
```
nit: apply python formatting there should be a space separating these functions

ardzoht commented on lte/gateway/python/integ_tests/s1aptests/s1ap_utils.py:
```diff
@@ -1508,6 +1513,20 @@ def get_gtp_port_no(self) -> Optional[int]:
+                return port_info[1]
+    # RYU rest API is not able dump flows from non zero table.
+    # this adds similar API using `ovs-ofctl` cmd
+    def get_flows(self, table_id) -> []:
```
```suggestion
    def get_flows(self, table_id) -> List[str]:
```

ardzoht commented on lte/gateway/python/integ_tests/s1aptests/s1ap_utils.py:
```diff
@@ -1524,3 +1543,51 @@ def offload_agw(self, imsi, enbID, offloadtype=0):
+
+        for conf in self.dump['configs']:
+            if 'dynamic_listeners' in conf:
+                return conf['dynamic_listeners'][0]['active_state']['listener']['filter_chains'][0]['filters']
```
can we confirm that those nested map and list always will be non-empty? this could throw `AttributeError` if the key or value doesn't exist

ardzoht reviewed (approved):
some minor nits, please take a look before merging, overall lgtm
